pkg/controller/node/node_controller.go

```diff
@@ -816,9 +949,12 @@ func (ctrl *Controller) syncMachineConfigPool(key string) error {
 	for _, node := range nodes {
 		// All the nodes that need to be upgraded should have `NodeUpdateInProgressTaint` so that they're less likely
 		// to be chosen during the scheduling cycle.
-		targetConfig := pool.Spec.Configuration.Name
+		// targetConfig := pool.Spec.Configuration.Name
 		hasInProgressTaint := checkIfNodeHasInProgressTaint(node)
-		if node.Annotations[daemonconsts.DesiredMachineConfigAnnotationKey] == targetConfig {
+
+		lns := ctrlcommon.NewLayeredNodeState(node)
+
+		if lns.IsDesiredEqualToPool(pool) {
```

Comment:
Just to understand, this check works for all pools layered or not, right?

Reply:
Yes it does. It first determines if it should consider images or not as it determines equality based upon whether the pool is opted into layering and whether it has an image or not.